Implement GATT_CL btp service

piotrnarajowski opened the pull request:
- Implement GATT Client (0x06) BTP service. Relevant components are adjusted accordingly to ensure proper integration. 
- Helper functions for all necessary GATT procedures were added to reduce amount of code in GATT test case handlers and to choose between GATT/GATT_CL BTP commands depending on which one is supported by IUT
- Implement necessary tools for event handling
- Implement read supported services btp command. This is used when initializing GATT CL service

piotrnarajowski committed:
btp: implement GATT Client (0x06) BTP service

This commit implements gatt_cl btp service. Relevant components are
adjusted accordingly to ensure proper integration. Helper functions
for all necessary GATT procedures were added to reduce amount of
code in GATT test case handlers and to choose between GATT/GATT_CL BTP
commands depending on which one is supported by IUT.

piotrnarajowski committed:
Add tools for event handling

This is needed for GATT_CL event handling.

piotrnarajowski commented:
@KKopyscinski please review

KKopyscinski reviewed (approved):
This is quite a big change, but I know this is ported from auto-pts, so I assume it's okay. Would be nice if you could run the test set and leave comment with results to leave a mark that this was working. Also, if something doesn't pass, please check if this is a regression. Otherwise, LGTM.

KKopyscinski commented on testcases/GattTestCase.py:
```diff
@@ -43,106 +44,137 @@ def test_btp_GATT_CL_GAD_1(self):
-
-        db = GattDB()
-        btp.gattc_disc_prim_svcs_rsp(self.iut1, db)
+        if stack.gatt_cl:
```
Maybe it is a good idea to abstract this logic as helper function, like so:
```suggestion
def gattc_disc_prim_svcs(addr1, addr2):
    if stack.gatt_cl:
            btp.gatt_cl_disc_prim_svcs(addr1, addr2)
            stack.gatt_cl.wait_for_prim_svcs()
            db = stack.gatt_cl.db
        else:
            btp.gattc_disc_prim_svcs(addr1, addr2)
            db = GattDB()
            btp.gattc_disc_prim_svcs_rsp(addr1, db)
    return db
```

and then, here do:
db = gattc_disc_prim_svcs(self.iut1, self.iut2.stack.gap.iut_addr_get())

applies to the rest of `if gatt_cl`, too

piotrnarajowski replied:
That's a great idea, will apply it

piotrnarajowski replied:
@KKopyscinski Helper functions for all GATT procedures are now implemented and used in GATT test cases.
All tests finished with PASS verdict except for one which for some reason randomly fails at connection procedure. I'm looking into this issue right now but this is unrelated to test case handlers that this PR changes

![image](https://github.com/user-attachments/assets/bcce72b5-c306-45cf-89ee-3f0c64649121)

KKopyscinski reviewed (changes requested):
I will mark this as changes requested so it won't be merged by accident before you're done

piotrnarajowski commented on testcases/utils.py:
```diff
@@ -25,13 +25,17 @@
 
 EV_TIMEOUT = 20
 
-
```
Fixed

piotrnarajowski committed:
btp: add read_supp_svcs command

For some reason implementation for this command was missing.
This will be used in preconditions to check if IUT supports
GATT CL BTP service.

KKopyscinski reviewed (approved):
Nice, looks much more cleaner now!
